File: apps/otel-test/src/instrumentation.ts
/**
 * OpenTelemetry Auto-Instrumentation
 *
 * Load before your application code:
 *   node --import ./src/instrumentation.ts src/index.ts
 *
 * Configuration via environment variables:
 *   OTEL_SERVICE_NAME            — service name (defaults to "otel-test")
 *   OTEL_EXPORTER_OTLP_ENDPOINT — OTLP backend URL (if not set, uses console exporter)
 *   OTEL_EXPORTER_OTLP_HEADERS  — auth headers for the backend
 *   OTEL_ENABLED_CATEGORIES     — comma-separated categories to export (default: all)
 */

import { NodeSDK } from "@opentelemetry/sdk-node";
import { getNodeAutoInstrumentations } from "@opentelemetry/auto-instrumentations-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { ConsoleSpanExporter } from "@opentelemetry/sdk-trace-base";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { ATTR_SERVICE_NAME } from "@opentelemetry/semantic-conventions";
import { FilteringExporter } from "./filtering-exporter";

const resource = resourceFromAttributes({
	[ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME ?? "otel-test",
});

// Use OTLP if endpoint configured, otherwise console
const traceExporter = process.env.OTEL_EXPORTER_OTLP_ENDPOINT
	? new FilteringExporter(new OTLPTraceExporter())
	: new ConsoleSpanExporter();

const sdk = new NodeSDK({
	resource,
	traceExporter,
	instrumentations: [
		getNodeAutoInstrumentations({
			"@opentelemetry/instrumentation-fs": { enabled: false },
			"@opentelemetry/instrumentation-dns": { enabled: false },
		}),
	],
});

sdk.start();
console.log(`OTel initialized: ${process.env.OTEL_EXPORTER_OTLP_ENDPOINT ? "OTLP" : "console"} exporter`);

process.on("SIGTERM", () => {
	sdk.shutdown().finally(() => process.exit(0));
});

File: apps/indusk-mcp/templates/instrumentation.py
"""
OpenTelemetry Auto-Instrumentation for Python

Run your application with the auto-instrumentation wrapper:

    opentelemetry-instrument python your_app.py

Or import this module before your application code:

    import instrumentation  # noqa: F401
    from your_app import main
    main()

Configuration via environment variables:
    OTEL_SERVICE_NAME              — service name (defaults to "unknown-service")
    OTEL_EXPORTER_OTLP_ENDPOINT   — OTLP backend URL
    OTEL_EXPORTER_OTLP_HEADERS    — auth headers (e.g., "Authorization=Bearer xxx")
    OTEL_PYTHON_LOG_CORRELATION   — set to "true" for trace context in logs (default)

Install required packages:
    pip install opentelemetry-distro opentelemetry-instrumentation opentelemetry-exporter-otlp

Then auto-install all available instrumentations:
    opentelemetry-bootstrap --action=install
"""

import os
import logging

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import (
    BatchSpanProcessor,
    ConsoleSpanExporter,
)
from opentelemetry.sdk.resources import Resource, SERVICE_NAME

# Configure resource with service name
resource = Resource.create(
    {SERVICE_NAME: os.environ.get("OTEL_SERVICE_NAME", "unknown-service")}
)

provider = TracerProvider(resource=resource)

# Use OTLP exporter if endpoint is configured, otherwise console
otlp_endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT")
if otlp_endpoint:
    from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
        OTLPSpanExporter,
    )

    provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
else:
    provider.add_span_processor(BatchSpanProcessor(ConsoleSpanExporter()))

trace.set_tracer_provider(provider)

# Enable trace context correlation in Python logging
os.environ.setdefault("OTEL_PYTHON_LOG_CORRELATION", "true")

# Configure structured logging
logging.basicConfig(
    level=os.environ.get("LOG_LEVEL", "INFO").upper(),
    format="%(asctime)s %(levelname)s [%(name)s] [trace_id=%(otelTraceID)s span_id=%(otelSpanID)s] %(message)s",
)
